File: solutions/elixir/135-inorder-successor-in-bst.ex
# Problem 135: Inorder Successor in BST (LeetCode 285)
# Difficulty: Med
# Language: Elixir
# 
defmodule Solution do
  def inorder_successor(nil, _p), do: nil
  def inorder_successor(%{val: v, left: left}, p) when v > p do
    case inorder_successor(left, p) do
      nil -> v
      res -> res
    end
  end
  def inorder_successor(%{right: right}, p), do: inorder_successor(right, p)
end

File: solutions/ocaml/135-inorder-successor-in-bst.ml
(*
  Problem 135: Inorder Successor in BST (LeetCode 285)
  Difficulty: Med
  Language: OCaml
*)
type tree = Leaf | Node of int * tree * tree

let rec inorder_successor root p =
  match root with
  | Leaf -> None
  | Node (v, left, _) when v > p ->
    (match inorder_successor left p with
     | None -> Some v
     | res  -> res)
  | Node (_, _, right) -> inorder_successor right p
